### ISSUE TITLE
Add support for a configuration file

### ISSUE THREAD

**Issue by marklonquist** [GoogleCloudPlatform/cloud-sql-proxy#1647]
## Feature Description
I propose a way to add instances through a file, rather than only through env variables and CLI args.
One way would be the following:

```go
filePath := os.Getenv(fmt.Sprintf("%s_INSTANCES_FILEPATH", envPrefix))
if filePath != "" {
  body, err := ioutil.ReadFile(filePath)
  if err != nil {
    logger.Errorf("error reading %s: %v", filePath, err)
    return nil
  }
  
  type InstancesFile struct {
    Instances []string `json:"instances"`
  }
  
  var instancesFile InstancesFile
  err = json.Unmarshal(body, &instancesFile)
  if err != nil {
    logger.Errorf("error unmarshalling %s: %v", filePath, err)
    return nil
  }
  args = append(args, instancesFile.Instances...)
}
```

and the docs could read:
```
It is also possible to use a file with the following syntax (JSON):

  {
    "instances": [
      "olympus-pub-dev-3d05:europe-west4:postgresql-01"
    ]
  }

where each entry to the array is a instance connection name.
```

This would make it possible to use zero-trust stores like hashicorp vault and load in through filesystem.

## Alternatives Considered
Haven't been able to locate any alternatives.

## Additional Context
We currently use a custom version of cloud_sql_proxy with the above code and is working brilliantly. I assume more could use this kind of feature, but time will tell :-)

**Comment by jackwotherspoon** [comment]
@marklonquist Thanks for raising a feature request on the proxy!

I think it is an interesting idea and will definitely discuss it with the team 😄 

In the meantime, I will monitor this issue to see if it gets any other interest from users of the proxy! 👍 

Thanks again and have an awesome day!

**Comment by enocom** [comment]
@marklonquist We're actually thinking about adding support for a configuration file for all arguments, instances included.

This opens the door to support dynamic configuration from remote sources (e.g., Secret Manager, Vault, etc).

**Comment by marklonquist** [comment]
@enocom this sounds awesome! Do you have any ETA on this? :)

**Comment by enocom** [comment]
:sweat_smile: Sorry it's dragging on. This has been pushed below other priorities. We still have this in our backlog.

We would welcome a PR for this, if you're open to making a few adjustments like:

- using TOML as a file format (or possibly YAML if TOML seems too unfamiliar)
- Matching the existing envvar config for both a single instance or multiple instances.

I put together a PR awhile back if it's a useful reference: https://github.com/GoogleCloudPlatform/cloud-sql-proxy/compare/main...config-file.


Otherwise, we'll move this up in our list.

**Comment by marklonquist** [comment]
I added a PR: #2106 :)

**Comment by enocom** [comment]
Awesome! I'll take a look soon.

**Comment by marklonquist** [comment]
I updated the PR to accept TOML, YAML and JSON config files. Is that OK? :)

**Comment by enocom** [comment]
Yeah, that's a great idea. As long as we support a format that includes comments, I think we're OK to support JSON too.

**Comment by enocom** [comment]
Fixed in #2106
